### ISSUE TITLE
How can I ensure MoltenVK compatibility when developing on Windows?

### ISSUE THREAD

**Issue by jjYBdx4IL** [KhronosGroup/MoltenVK#1533]
Hi!

I want to make sure that I can port my project from Windows to Mac/MoltenVK without much additional hassle. Is there any recommneded way for doing this? Any MoltenVK validation that I can activate on Windows - apart from not specifying a Vulkan API higher than 1.1?

TY!

**Comment by MennoVink** [comment]
Aside from the core version, also check if mvk supports an extension before you plan on using it on windows. Besides that it's probably best to develop both platforms simultaneously, if not macos first. mvk has more limitations and caveats than native windows drivers, so if you make sure it works with mvk then it'll work on windows too. Doing it the other way around may result in quite a bit of 'additional hassle' trying to get macos to perform somewhat on-par with the windows implementation.
Some things that caused me quite a bit of hassle:
- SwapChain synchronization is broken with occluded windows. If you're doing rendering on another thread than the main one this is going to be an issue. Dont try and be smart trying to work around it, just use a CVDisplayLink.
- Transcribing from vulkan to metal is slow ... and i mean real slow. The only workaround for this is to move transcribing from submission to recording. You must then record/transcribe commands in parallel to mitigate this issue. Note that command buffers cannot be reused either.
- Lack of some basic extensions, ie even ones not requiring specific hardware like sync2 and dynamic rendering are missing. So yeah definitely check the mvk supported extensions list before making any plans.

Depending on the size of your project the best option might be to abstract the rendering hardware similar to how the big game engines are doing. You send a 'scene description' to the renderer and the renderer's implementation takes care of translating that to native graphics api calls. If that's an option however then you probably wont be using vulkan on win/macos but yeah just leaving that out there.

**Comment by jjYBdx4IL** [comment]
ty for the valuable input

**Comment by richard-lunarg** [comment]
As MennoVink suggested, I typically develop on both simultaneously. There
is a new validation layer and tools set that is part of the Vulkan SDK that
may be of use to you as well,

https://www.lunarg.com/wp-content/uploads/2022/03/The-Vulkan-Profiles-Toolset-Solution-FEB2022.pdf

The profiles library lets you easily select a set of extensions that are
known good on another platform, and you can validate against the profiles
layer to ensure you stay within the limits of a given profile. New profiles
will be available as time progresses. Check out the link above for more
information.


Richard S. Wright Jr.
Sr. Graphics Software Engineer
LunarG, Inc. - 3D Graphics Software Solutions
***@***.***

**Comment by billhollings** [comment]
@MennoVink Thanks for feedback and insight. It's very helpful.

I encourage you (and all MoltenVK users) to post Issues for specific concerns, improvement suggestions, and enhancement requests. That kind of information from active users is very useful to help us prioritize MoltenVK development.

Specifically, feel free to identify specific missing Vulkan extensions that we can add that would be useful to you! Vulkan has a large ever-growing list of extensions, and knowing which are most useful to users would help us prioritize how we add them.

**Comment by MennoVink** [comment]
@billhollings 
The swapchain synchronization issue i havn't reported here as it's not something that's mvk/metal specific. We've had this issue with opengl too on macos. I'm actually quite happy this issue exists, because it pushed us to use displaylinks which assist with smoothness :). I think this area is where windows is lacking behind. Linux has the presentation timers that could be used but the support for presentation timing extensions on windows isn't there.

For the transcribing overhead i dont have enough metal knowledge to know what to ask for, or if it's even possible to get this to be faster. A lot of time is spent in actual metal calls so the only way to change this if mvk makes different calls, but yeah if those dont exist in metal... I had created [this issue](https://github.com/KhronosGroup/MoltenVK/issues/1473) back when i first ran into this problem, but since timeline semaphores cant be used to synchronize more than 1 producer then that doesn't solve it. The only way to bypass this is to use prefillMetalCommandBuffers with parallel recording. What bothers me here is the maxActiveMetalCommandBuffersPerQueue configuration value. This is one of those macos specific limitations. My engine doesn't know how many command buffers there are as it's affected by how much stuff the user adds to the scene and os thread scheduling. This is fine on windows, just create a pool per thread per frame and done, on macos however i dont know. I've already ran into freezes when this value was too low, currently i've guestimated it could be up to 2k so i've filled that in and no freezes anymore, but i have no idea what 'normal' numbers should be here or what raising them does (i gues memory usage?) Do you think it's useful to get a new issue for this?

The lack for basic extensions is a larger issue imo. For sync2 specifically i've made https://github.com/KhronosGroup/MoltenVK/issues/1514 quite a while back already. For me, long enough back to decide that requesting a feature when i need it for our next release is not the way to go. I've since had to do the translation back to sync 1 manually. What worries me most on this front is how mvk is falling behind more and more and how the insights in number of commits actually seems to be going down instead of up. It seems to have dropped to 50% of what it was before july/2021. Commit velocity is not the greatest indicator of progress obviously. Large projects like the mvk metal extension could take up a lot of time, but that doesn't take away that we're now 2 core versions and ~100 extensions behind. I've mentioned this during the vulkan 1.3 webinar. @KarenGhavam-lunarG aknowledged that this issue is recognised, but also mentioned there are no concrete plans yet on having mvk catch up. If you need priorities i gues catching up with 1.3 core would be a good guideline.

I realize that feedback like this might sound overly negative, so i want to take this opportunity to say thank you again for making/maintaining mvk. I realize it is a lot of work and i really appreciate the effort. I think Vulkan is awsome and being able to do it in a cross-platform manner is well worth a minor headache every now and then :)

**Comment by jjYBdx4IL** [comment]
Thx for the input. I guess running NanoVG over MoltenVK seems to be a particularly bad idea because of the transcribe over head.
